### PR TITLE
Node Provider Support

### PR DESCRIPTION
**Pull request by dahendel**
This PR provides a `providers` interface via the providers package to enable node configuration from other sources ie docker-machine. A docker-machine provider is also provided. README.md has been updated with documentation as well.

Replaces #442  

@moelsayed @superseb

### CODE REVIEW

**Comment by deniseschannon** [comment]
https://github.com/rancher/rke/issues/429

File: cmd/config.go [hunk 1]
```diff
@@ -94,30 +100,60 @@ func clusterConfig(ctx *cli.Context) error {
...
+	cluster.SSHKeyPath = sshKeyPath
+
 	cluster.Nodes = make([]v3.RKEConfigNode, 0)
-	for i := 0; i < numberOfHostsInt; i++ {
-		hostCfg, err := getHostConfig(reader, i, cluster.SSHKeyPath)
+	if nodeProvider != nil {
+
+		machines, err := nodeProvider.ListNodes(reader)
+
+		if machines[0] == "" {
```
Comment (moelsayed) [inline cmd/config.go]: Why no use `len()` ?

File: providers/docker-machine.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		} `json:"EngineOptions,omitempty"`
+	}
+	Name string `json:"Name,omitempty"`
+}
+
+func (d *DockerMachineProvider) ListNodes(reader *bufio.Reader) ([]string, error) {
+
+	var err error
+
+	// Set the default Docker Machine Storage Path
```
Comment (moelsayed) [inline providers/docker-machine.go]: Can you please move the docker machine path logic to it's own function ?

File: providers/docker-machine.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	} `json:"Driver,omitempty"`
+	HostOptions struct {
+		EngineOptions struct {
+			Labels []string `json:"Labels,omitempty"`
+		} `json:"EngineOptions,omitempty"`
+	}
+	Name string `json:"Name,omitempty"`
+}
+
+func (d *DockerMachineProvider) ListNodes(reader *bufio.Reader) ([]string, error) {
```
Comment (moelsayed) [inline providers/docker-machine.go]: The function doesn't exactly list nodes.. The name is a bit confusing. Can you please rename it to something more descriptive?

File: providers/provider-types.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"strings"
+	"sync"
+)
+
+type NodeProvider interface {
```
Comment (alena1108) [inline providers/provider-types.go]: Lets rename it to NodeConfig provider to be clear that it's not actually provisioning the node, but just returning the config. Readme has to be changed accordingly

File: providers/provider-types.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	ListNodes(reader *bufio.Reader) ([]string, error)
+	ReadNodeConfigurations(deployNodes []string) ([]v3.RKEConfigNode, error)
```
Comment (alena1108) [inline providers/provider-types.go]: Lets add Register to the list of methods. Things like setting up environment variables have be done at registration point, and it should fail if something goes wrong

File: providers/provider-types.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+var (
+	nodeProviders = map[string]NodeProvider{
+		"docker-machine": new(DockerMachineProvider),
+	}
+	nodeProvidersMu = sync.RWMutex{}
+)
+
+func SetNodeProvider(name string, provider NodeProvider) {
+	nodeProvidersMu.Lock()
```
Comment (alena1108) [inline providers/provider-types.go]: I don't believe this method is ever called

File: providers/provider-types.go (hunk 5 above)
Comment (alena1108) [reply providers/provider-types.go]: You can look at these code bases as an example of providers registration: https://github.com/rancher/lb-controller/blob/master/provider/provider.go or https://github.com/rancher/external-dns/blob/master/providers/provider.go

**Comment by moelsayed** [comment]
@dahendel Can you please rename the files without the dashes, and create a subdirectory for each provider ?

File: README.md [hunk 7]
```diff
@@ -241,6 +241,33 @@ rke config --name mycluster.yml
...
+Node Providers are implemented through the NodeProvider interface in the providers package.
+
+Available node providers:
+- docker-machine
+
+##### Docker-Machine Provider
+RKE will prompt for the docker-machine store-path (/home/$USER/.docker/machine/machines) in case a custom path was used when creating the machines. The next step is to provide a comma separated list of the machines to use for the nodes.
+It is important to note that the  user used during creation with docker-machine, which gets written to cluster.yml, MUST be in the docker group on the machine, or rke will fail to connect.
+
+This option also assumes that labels were passed to the engine and reads them to configure the node:
```
Comment (moelsayed) [inline README.md]: it should be clear that the engine is docker-machine

**Comment by dahendel** [comment]
@moelsayed @deniseschannon Thanks for the feedback. I have been out the past week, so I will get caught up and then start on these changes.

**Comment by dahendel** [comment]
@moelsayed @deniseschannon Got this updated. Any other thoughts?

**Comment by moelsayed** [comment]
@dahendel thank you for the updates, can you please rebase?

**Comment by moelsayed** [comment]
@dahendel also, please remove the [binary](https://github.com/rancher/rke/pull/600/files?utf8=%E2%9C%93&diff=split#diff-1af5c5571bf827029b33d30baa03d8d6)

**Comment by dahendel** [comment]
Will do.

**Comment by dahendel** [comment]
@moelsayed Where should I document the Node Providers documentation now?

**Comment by moelsayed** [comment]
@dahendel Thank you so much for the contribution. Could you please rebase to the latest master ?

**Comment by dahendel** [comment]
@moelsayed I will try and get that knocked out tonight or tomorrow.

**Comment by moelsayed** [comment]
LGTM, thank you!

**Comment by catherineluse** [comment]
@cjellick There's a PR about this for the docs https://github.com/rancher/docs/pull/842 that can't be merged because it depends on this being merged. is this going to be merged?

**Pull request closed without merging**